### PR TITLE
[13.0][ADD] purchase_order_product_summary

### PR DESCRIPTION
**Pull request by ChristianSantamaria**
Add a new section with the grouped products and their total quantity.

### CODE REVIEW

File: purchase_order_product_summary/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    "summary": """
+        Add a new section with the grouped products and their total quantity. 
+    """,
+    "author": "Solvos",
+    "license": "LGPL-3",
+    "version": "13.0.1.0.0",
+    "category": "stock",
+    "website": "https://github.com/solvosci/slv-stock",
+    "depends": [
+        "purchase"
```
Comment (dalonsod) [inline purchase_order_product_summary/__manifest__.py]: Suggested change:
```suggestion
        "purchase",
```

File: purchase_order_product_summary/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    ],
+    "data": [
+        "report/purchase_order_summary_template.xml"
```
Comment (dalonsod) [inline purchase_order_product_summary/__manifest__.py]: Suggested:
```suggestion
        "report/purchase_order_summary_template.xml",
```

File: purchase_order_product_summary/report/purchase_order_summary_template.xml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <template id="report_purchaseorder_document_product_summary" inherit_id="purchase.report_purchaseorder_document">
+        <xpath expr="//div[@id='total']" position="after">
+            <div class="row justify-content-start">
+                <div class="col-4">
+                    <h4>Summary Totals:</h4>
```
Comment (dalonsod) [inline purchase_order_product_summary/report/purchase_order_summary_template.xml]: Better this? I'm not sure
```suggestion
                    <h4>Product Summary:</h4>
```

File: purchase_order_product_summary/report/purchase_order_summary_template.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                </div>
+            </div>
+            <table class="table table-sm o_main_table">
+                <thead>
+                    <tr>
+                        <th name="th_id"><strong>Code</strong></th>
+                        <th name="th_name"><strong>Product Name</strong></th>
```
Comment (dalonsod) [inline purchase_order_product_summary/report/purchase_order_summary_template.xml]: @lmiguens-solvos what do you prefer? Separed columns, or both code and product in the same column?

File: purchase_order_product_summary/report/purchase_order_summary_template.xml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+                        <th name="th_summary" class="text-right"><strong>Total</strong></th>
+                    </tr>
+                </thead>
+                <tbody>
+                    <t t-foreach="o.order_line.mapped('product_id')" t-as="line">
```
Comment (dalonsod) [inline purchase_order_product_summary/report/purchase_order_summary_template.xml]: Please rename, `line` is ambiguous. Suggested:
```suggestion
                    <t t-foreach="o.order_line.mapped('product_id')" t-as="product_agg_line">
```

File: purchase_order_product_summary/report/purchase_order_summary_template.xml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+                                <span t-esc="sum(o.order_line.filtered(lambda x: x.product_id.id == line.id).mapped('product_qty'))" t-options='{"widget": "float", "precision": 3}'/>
+                                <span t-esc="line.uom_id.name" groups="uom.group_uom"/>
+                            </td>
+                        </tr>
+                    </t>
+                </tbody>
+            </table>
+        </xpath>
+    </template>
+</odoo>
```
Comment (dalonsod) [inline purchase_order_product_summary/report/purchase_order_summary_template.xml]: Missing EOF line

File: purchase_order_product_summary/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# © 2021 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Purchase Order Product Summary",
+    "summary": """
+        Add a new section with the grouped products and their total quantity.
```
Comment (dalonsod) [inline purchase_order_product_summary/__manifest__.py]: Incomplete summary IMO:
```suggestion
        Adds to purchase report a new section with the grouped products and their total quantity.
```

**Review by dalonsod — CHANGES_REQUESTED** [review]
See comments. Only code review, I'll test it later

File: purchase_order_product_summary/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Purchase Order Product Summary",
+    "summary": """
+        Add a new section with the grouped products and their total quantity. 
+    """,
+    "author": "Solvos",
+    "license": "LGPL-3",
+    "version": "13.0.1.0.0",
+    "category": "stock",
```
Comment (dalonsod) [inline purchase_order_product_summary/__manifest__.py]: Wrong category. See https://github.com/odoo/odoo/blob/13.0/addons/purchase/__manifest__.py#L7

File: purchase_order_product_summary/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# © 2021 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Purchase Order Product Summary",
+    "summary": """
+        Adds to purchase report a new section with the grouped products and their total.
```
Comment (dalonsod) [inline purchase_order_product_summary/__manifest__.py]: Line too long (> 80)

File: purchase_order_product_summary/report/purchase_order_summary_report.xml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
+        <report 
+            string="Product Summary"
```
Comment (dalonsod) [inline purchase_order_product_summary/report/purchase_order_summary_report.xml]: I think better named "Purchase order with product summary"

File: purchase_order_product_summary/report/purchase_order_summary_report.xml [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+<odoo>
+    <data>
+        <report 
+            string="Product Summary" 
+            id="action_report_purchase_order" 
+            model="purchase.order" 
+            report_type="qweb-pdf"
+            name="purchase_order_product_summary.report_purchasesummary" 
+            file="purchase_order_product_summary.report_purchasesummary"
+            print_report_name="'Product Summary - %s' % (object.name)"
```
Comment (dalonsod) [inline purchase_order_product_summary/report/purchase_order_summary_report.xml]: The same as above

File: purchase_order_product_summary/report/purchase_order_summary_template.xml [hunk 12]
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <template id="report_purchasesummary_document">
+        <div class="page">
```
Comment (dalonsod) [inline purchase_order_product_summary/report/purchase_order_summary_template.xml]: For further inheritances:
```suggestion
        <div class="page" id="product_summary">
```

**Review by dalonsod — CHANGES_REQUESTED** [review]
Good work, almost finished. See comments

File: purchase_order_product_summary/__manifest__.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
+# © 2021 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Purchase Order Product Summary",
+    "summary": """
+        Adds to purchase report a new section \n
```
Comment (dalonsod) [inline purchase_order_product_summary/__manifest__.py]: I think this is not necessary:
```suggestion
        Adds to purchase report a new section
```

File: purchase_order_product_summary/report/purchase_order_summary_template.xml (hunk 4 above)
Comment (lmiguens-solvos) [reply purchase_order_product_summary/report/purchase_order_summary_template.xml]: I really don't care. 
I don't really know what will be better. Seeing that the product codes are all the same length ... well, it doesn't matter.

**Review by dalonsod — CHANGES_REQUESTED** [review]
See comment. Also, maybe .pot and .po files should be re-created

File: purchase_order_product_summary/report/purchase_order_summary_template.xml (hunk 4 above)
Comment (dalonsod) [reply purchase_order_product_summary/report/purchase_order_summary_template.xml]: got it :)

**Review by dalonsod — APPROVED** [review]

**Comment by dalonsod** [comment]
@lmiguens-solvos is it OK for you?

**Comment by lmiguens-solvos** [comment]
> @lmiguens-solvos is it OK for you?

The block with the summary seems to me to be perfect. The only thing that would be fine is to start on a new page (to print only that summary)
The detailed block (for me) I would remove the date column and the long description of the VAT type (which could be hidden or just put the code) ... but hey, at the moment we don't do anything, we wait for what the client says.
Another thing, ... Although I have the user in Spanish, and I loaded translations, I get the format in English ... ??

**Comment by dalonsod** [comment]
> The only thing that would be fine is to start on a new page (to print only that summary)

I agree. @ChristianSantamaria could you change it?



> The detailed block (for me) I would remove the date column and the long description of the VAT type (which could be hidden or just put the code) ... but hey, at the moment we don't do anything, we wait for what the client says.

That would be a genuine custom format, and could be developed in another addon (nothing related to product summary concept), but not right now, as you said

> Another thing, ... Although I have the user in Spanish, and I loaded translations, I get the format in English ... ??

Odoo applies `partner_id` language for purchase documents by default, and this addon works in the same way. So, if you change the partner lang, for the purchase selected partner, you'll get report in the desired language.

**Comment by lmiguens-solvos** [comment]
> Odoo applies `partner_id` language for purchase documents by default, and this addon works in the same way. So, if you change the partner lang, for the purchase selected partner, you'll get report in the desired language.

You're right. Tried and ok (despite "Monto")

**Comment by dalonsod** [comment]
>You're right. Tried and ok (despite "Monto")

Well, LATAM already rules in Odoo spanish world...

**Comment by ChristianSantamaria** [comment]
> The only thing that would be fine is to start on a new page (to print only that summary)

Done, now product summary is on a new page

**Comment by lmiguens-solvos** [comment]
> > The only thing that would be fine is to start on a new page (to print only that summary)
> 
> Done, now product summary is on a new page

Viewed! Well but thinking that they will want to print this sheet separately, I am missing something of the header: purchase order and date of the order, for example. I don't think the whole header of the detailed format is needed. 
What do yo think @dalonsod ??

File: purchase_order_product_summary/report/purchase_order_summary_template.xml [hunk 14]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <template id="report_purchasesummary_document">
+        <t t-call="web.external_layout">
+            <t t-set="o" t-value="o.with_context(lang=o.partner_id.lang)"/>
+            <div class="page" id="product_summary">
+                <div class="oe_structure"/>
+                <div class="row justify-content-start">
```
Comment (dalonsod) [inline purchase_order_product_summary/report/purchase_order_summary_template.xml]: > I am missing something of the header: purchase order and date of the order, for example. I don't think the whole header of the detailed format is needed.

Product summary was implemented as an independent template, so we need to add again those fields. Then, no problem adding the fields we're interested in. This works:

```suggestion
                <div class="row justify-content-start">
                <h2 t-if="o.state == 'draft'">Request for Quotation #<span t-field="o.name"/></h2>
                <h2 t-if="o.state in ['sent', 'to approve']">Purchase Order #<span t-field="o.name"/></h2>
                
                <div id="summary_informations" class="row mt32 mb32">
                  <div t-if="o.date_order" class="col-3 bm-2">
                      <strong>Order Date:</strong>
                      <p t-field="o.date_order" class="m-0"/>
                  </div>
                </div>
```

File: purchase_order_product_summary/report/purchase_order_summary_template.xml [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+                    <div class="col-4">
+                        <h4>Product Summary:</h4>
+                    </div>
+                </div>
+                <table class="table table-sm">
```
Comment (dalonsod) [inline purchase_order_product_summary/report/purchase_order_summary_template.xml]: For further inheritance:

```suggestion
                <table class="table table-sm" id="summary_table">
```

**Review by dalonsod — CHANGES_REQUESTED** [review]
See comments. Suggested changes will need i18n files update.

**Review by dalonsod — APPROVED** [review]
LGTM
Please squash commits, and I'll merge it, if @lmiguens-solvos approves

**Review by lmiguens-solvos — APPROVED** [review]
LGTM

**Comment by dalonsod** [comment]
Merge is coming

**Pull request merged**